tests/functional/builtins/codegen/test_abi_decode_fuzz.py

```diff
@@ -0,0 +1,371 @@
+from dataclasses import asdict, dataclass
+
+import hypothesis as hp
+import hypothesis.strategies as st
+import pytest
+from eth.codecs import abi
+
+from tests.evm_backends.base_env import EvmError
+from vyper.codegen.core import calculate_type_for_external_return, needs_external_call_wrap
+from vyper.semantics.types import (
+    AddressT,
+    BoolT,
+    BytesM_T,
+    BytesT,
+    DArrayT,
+    DecimalT,
+    HashMapT,
+    IntegerT,
+    SArrayT,
+    StringT,
+    TupleT,
+    VyperType,
+    _get_primitive_types,
+    _get_sequence_types,
+)
+from vyper.semantics.types.shortcuts import UINT256_T
+
+from .abi_decode import DecodeError, spec_decode
+
+pytestmark = pytest.mark.fuzzing
+
+type_ctors = []
+for t in _get_primitive_types().values():
+    if t == HashMapT or t == DecimalT():
+        continue
+    if isinstance(t, VyperType):
+        t = t.__class__
+    if t in type_ctors:
+        continue
+    type_ctors.append(t)
+
+complex_static_ctors = [SArrayT, TupleT]
+complex_dynamic_ctors = [DArrayT]
+leaf_ctors = [t for t in type_ctors if t not in _get_sequence_types().values()]
+static_leaf_ctors = [t for t in leaf_ctors if t._is_prim_word]
+dynamic_leaf_ctors = [BytesT, StringT]
+
+MAX_MUTATIONS = 33
```

Comment:
what is the reasoning behind using 33?
why is it ok to use 33 for all payloads, irrespective of their lengths? ie why do we assume this works well both for short and long payloads?

Reply:
It's about editing, you can add or delete one word (plus change)

Reply:
we bias towards the `w` action, which is word-level

Reply:
overall, i'm not sure about the byte-level mutation. if it happens on the `length` word, it will likely create an invalid length. if it happens on a pointer, it will likely create an invalid pointer (eg points to a too large address)

given the fairly low `MAX_MUTATIONS` if a bad byte mutation happens, I think there's a low probability of offsetting it with a "good" mutation

Reply:
like normal fuzzers do it on this level, but i think that the mutation depth there is much higher

Reply:
i think it's mostly useful for editing the "data" portion of the payload, like it produces off-by-ones (expected length is 32 but data is only actually 31)